In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
plt.style.use('dark_background')
import scipy

In [2]:
path = r'C:\Users\honoc\Documents\CareerFoundry Folder\Advanced Analytics and Dashboard Design\Zomato Folder'
path
df_zomato = pd.read_csv(os.path.join(path, 'Original data', 'zomato.csv'), index_col = False)

In [3]:
df_zomato.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

In [4]:
df_zomato.shape

(51717, 17)

In [5]:
df_zomato.columns

Index(['url', 'address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost(for two people)', 'reviews_list', 'menu_item',
       'listed_in(type)', 'listed_in(city)'],
      dtype='object')

Dropping columns not important for this project

In [6]:
df_zomato = df_zomato.drop(['url', 'address', 'phone', 'menu_item', 'dish_liked', 'reviews_list'], axis=1)

In [7]:
df_zomato.head()

name online_order book_table   rate  votes      location  \
0                  Jalsa          Yes        Yes  4.1/5    775  Banashankari   
1         Spice Elephant          Yes         No  4.1/5    787  Banashankari   
2        San Churro Cafe          Yes         No  3.8/5    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No  3.7/5     88  Banashankari   
4          Grand Village           No         No  3.8/5    166  Basavanagudi   

             rest_type                        cuisines  \
0        Casual Dining  North Indian, Mughlai, Chinese   
1        Casual Dining     Chinese, North Indian, Thai   
2  Cafe, Casual Dining          Cafe, Mexican, Italian   
3          Quick Bites      South Indian, North Indian   
4        Casual Dining        North Indian, Rajasthani   

  approx_cost(for two people) listed_in(type) listed_in(city)  
0                         800          Buffet    Banashankari  
1                         800          Buffet    Banashankari  
2                         800          Buffet    Banashankari  
3                         300          Buffet    Banashankari  
4                         600          Buffet    Banashankari

In [8]:
df_zomato.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   name                         51717 non-null  object
 1   online_order                 51717 non-null  object
 2   book_table                   51717 non-null  object
 3   rate                         43942 non-null  object
 4   votes                        51717 non-null  int64 
 5   location                     51696 non-null  object
 6   rest_type                    51490 non-null  object
 7   cuisines                     51672 non-null  object
 8   approx_cost(for two people)  51371 non-null  object
 9   listed_in(type)              51717 non-null  object
 10  listed_in(city)              51717 non-null  object
dtypes: int64(1), object(10)
memory usage: 4.3+ MB


Checking for Mixed-type columns

In [9]:
for col in df_zomato.columns.tolist():
    weird = (df_zomato[[col]].applymap(type) != df_zomato[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_zomato[weird]) > 0:
        print(col)

rate
location
rest_type
cuisines
approx_cost(for two people)


Checking for and Dropping Duplicates

In [10]:
df_zomato.drop_duplicates(inplace= True)

In [11]:
df_zomato.shape

(51609, 11)

108 DUPLICATES DROPPED

CLEANING UP THE MIXED-TYPE COLUMNS

STARTING WITH THE "rate" column. The "rate" is displayed as a fraction of 5. It has to be changed to a simple float value. That is, instead of 4.1/5, it should be 4.1. It could also possibly have other mixed-data types.

In [12]:
df_zomato['rate'].unique()

array(['4.1/5', '3.8/5', '3.7/5', '3.6/5', '4.6/5', '4.0/5', '4.2/5',
       '3.9/5', '3.1/5', '3.0/5', '3.2/5', '3.3/5', '2.8/5', '4.4/5',
       '4.3/5', 'NEW', '2.9/5', '3.5/5', nan, '2.6/5', '3.8 /5', '3.4/5',
       '4.5/5', '2.5/5', '2.7/5', '4.7/5', '2.4/5', '2.2/5', '2.3/5',
       '3.4 /5', '-', '3.6 /5', '4.8/5', '3.9 /5', '4.2 /5', '4.0 /5',
       '4.1 /5', '3.7 /5', '3.1 /5', '2.9 /5', '3.3 /5', '2.8 /5',
       '3.5 /5', '2.7 /5', '2.5 /5', '3.2 /5', '2.6 /5', '4.5 /5',
       '4.3 /5', '4.4 /5', '4.9/5', '2.1/5', '2.0/5', '1.8/5', '4.6 /5',
       '4.9 /5', '3.0 /5', '4.8 /5', '2.3 /5', '4.7 /5', '2.4 /5',
       '2.1 /5', '2.2 /5', '2.0 /5', '1.8 /5'], dtype=object)

Upon inspection, the "rate" column includes string values like 'NEW', '-', in addition to '/5'
Removing "NEW", "-", and "/5" from "rate" column

In [13]:
def cleanrate(value):
    if(value=='NEW' or value=='-'):
        return np.nan
    else:
        value = str(value).split('/')
        value = value[0]
        return float(value)

In [15]:
df_zomato['rate'] = df_zomato['rate'].apply(cleanrate)
df_zomato.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

             rest_type                        cuisines  \
0        Casual Dining  North Indian, Mughlai, Chinese   
1        Casual Dining     Chinese, North Indian, Thai   
2  Cafe, Casual Dining          Cafe, Mexican, Italian   
3          Quick Bites      South Indian, North Indian   
4        Casual Dining        North Indian, Rajasthani   

  approx_cost(for two people) listed_in(type) listed_in(city)  
0                         800          Buffet    Banashankari  
1                         800          Buffet    Banashankari  
2                         800          Buffet    Banashankari  
3                         300          Buffet    Banashankari  
4                         600          Buffet    Banashankari

Checking for null values in "rate" column

In [16]:
df_zomato.rate.isnull().sum()

10019

Filling Null Values in "rate" Column with Mean

In [18]:
df_zomato['rate'].fillna(df_zomato['rate'].mean(), inplace= True)
df_zomato['rate'].isnull().sum()

0

In [19]:
for col in df_zomato.columns.tolist():
    weird = (df_zomato[[col]].applymap(type) != df_zomato[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_zomato[weird]) > 0:
        print(col)

location
rest_type
cuisines
approx_cost(for two people)


In [20]:
df_zomato.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51609 entries, 0 to 51716
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   name                         51609 non-null  object 
 1   online_order                 51609 non-null  object 
 2   book_table                   51609 non-null  object 
 3   rate                         51609 non-null  float64
 4   votes                        51609 non-null  int64  
 5   location                     51588 non-null  object 
 6   rest_type                    51382 non-null  object 
 7   cuisines                     51564 non-null  object 
 8   approx_cost(for two people)  51265 non-null  object 
 9   listed_in(type)              51609 non-null  object 
 10  listed_in(city)              51609 non-null  object 
dtypes: float64(1), int64(1), object(9)
memory usage: 4.7+ MB


Dropping Null Values for "location", "rest_type", "cuisines", and "approx_cost(for two people)". Reason: The Null Values are few and will not have a significant effect on the outcome of the Analysis.

In [22]:
df_zomato.dropna(inplace= True)
df_zomato.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

             rest_type                        cuisines  \
0        Casual Dining  North Indian, Mughlai, Chinese   
1        Casual Dining     Chinese, North Indian, Thai   
2  Cafe, Casual Dining          Cafe, Mexican, Italian   
3          Quick Bites      South Indian, North Indian   
4        Casual Dining        North Indian, Rajasthani   

  approx_cost(for two people) listed_in(type) listed_in(city)  
0                         800          Buffet    Banashankari  
1                         800          Buffet    Banashankari  
2                         800          Buffet    Banashankari  
3                         300          Buffet    Banashankari  
4                         600          Buffet    Banashankari

Renaming columns: "approx_cost(for two people)" to "Cost2plates", "listed_in(type)" to "Type"

In [24]:
df_zomato.rename(columns = {'approx_cost(for two people)':'Cost2plates', 'listed_in(type)':'Type'}, inplace = True)
df_zomato.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

             rest_type                        cuisines Cost2plates    Type  \
0        Casual Dining  North Indian, Mughlai, Chinese         800  Buffet   
1        Casual Dining     Chinese, North Indian, Thai         800  Buffet   
2  Cafe, Casual Dining          Cafe, Mexican, Italian         800  Buffet   
3          Quick Bites      South Indian, North Indian         300  Buffet   
4        Casual Dining        North Indian, Rajasthani         600  Buffet   

  listed_in(city)  
0    Banashankari  
1    Banashankari  
2    Banashankari  
3    Banashankari  
4    Banashankari

Correcting for "location" and "listed_in(city)", both share same information. Therefore, one should be dropped to keep the dataframe tidier.

In [29]:
df_zomato = df_zomato.drop(['listed_in(city)'], axis= 1)

KeyError: "['listed_in(city)'] not found in axis"

In [30]:
df_zomato.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51042 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          51042 non-null  object 
 1   online_order  51042 non-null  object 
 2   book_table    51042 non-null  object 
 3   rate          51042 non-null  float64
 4   votes         51042 non-null  int64  
 5   location      51042 non-null  object 
 6   rest_type     51042 non-null  object 
 7   cuisines      51042 non-null  object 
 8   Cost2plates   51042 non-null  object 
 9   Type          51042 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 4.3+ MB


In [31]:
df_zomato.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

             rest_type                        cuisines Cost2plates    Type  
0        Casual Dining  North Indian, Mughlai, Chinese         800  Buffet  
1        Casual Dining     Chinese, North Indian, Thai         800  Buffet  
2  Cafe, Casual Dining          Cafe, Mexican, Italian         800  Buffet  
3          Quick Bites      South Indian, North Indian         300  Buffet  
4        Casual Dining        North Indian, Rajasthani         600  Buffet

Let's look at the Unique values in "Cost2plates"

In [36]:
df_zomato['Cost2plates'].unique()

array(['800', '300', '600', '700', '550', '500', '450', '650', '400',
       '900', '200', '750', '150', '850', '100', '1,200', '350', '250',
       '950', '1,000', '1,500', '1,300', '199', '80', '1,100', '160',
       '1,600', '230', '130', '50', '190', '1,700', '1,400', '180',
       '1,350', '2,200', '2,000', '1,800', '1,900', '330', '2,500',
       '2,100', '3,000', '2,800', '3,400', '40', '1,250', '3,500',
       '4,000', '2,400', '2,600', '120', '1,450', '469', '70', '3,200',
       '60', '560', '240', '360', '6,000', '1,050', '2,300', '4,100',
       '5,000', '3,700', '1,650', '2,700', '4,500', '140'], dtype=object)

There are some values in "Cost2plates" with commas, and these restrict it the values from becoming integers or floats. So let's convert that data types to integer. 

In [40]:
import pandas as pd

def clean_cost2plates(value):
    try:
        # Remove commas and convert to integer
        cleaned_value = int(value.replace(',', ''))
        return cleaned_value
    except ValueError:
        # Handle cases where conversion to integer is not possible
        return None

# Assuming df_zomato is your DataFrame
df_zomato['Cost2plates'] = df_zomato['Cost2plates'].apply(clean_cost2plates)


In [41]:
df_zomato['Cost2plates'].unique()

array([ 800,  300,  600,  700,  550,  500,  450,  650,  400,  900,  200,
        750,  150,  850,  100, 1200,  350,  250,  950, 1000, 1500, 1300,
        199,   80, 1100,  160, 1600,  230,  130,   50,  190, 1700, 1400,
        180, 1350, 2200, 2000, 1800, 1900,  330, 2500, 2100, 3000, 2800,
       3400,   40, 1250, 3500, 4000, 2400, 2600,  120, 1450,  469,   70,
       3200,   60,  560,  240,  360, 6000, 1050, 2300, 4100, 5000, 3700,
       1650, 2700, 4500,  140], dtype=int64)

In [42]:
for col in df_zomato.columns.tolist():
    weird = (df_zomato[[col]].applymap(type) != df_zomato[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_zomato[weird]) > 0:
        print(col)

In [43]:
df_zomato.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

             rest_type                        cuisines  Cost2plates    Type  
0        Casual Dining  North Indian, Mughlai, Chinese          800  Buffet  
1        Casual Dining     Chinese, North Indian, Thai          800  Buffet  
2  Cafe, Casual Dining          Cafe, Mexican, Italian          800  Buffet  
3          Quick Bites      South Indian, North Indian          300  Buffet  
4        Casual Dining        North Indian, Rajasthani          600  Buffet

Cleaning Rest Type column

In [44]:
df_zomato['rest_type'].value_counts()

rest_type
Quick Bites                   19010
Casual Dining                 10253
Cafe                           3682
Delivery                       2574
Dessert Parlor                 2242
                              ...  
Dessert Parlor, Kiosk             2
Food Court, Beverage Shop         2
Dessert Parlor, Food Court        2
Quick Bites, Kiosk                1
Sweet Shop, Dessert Parlor        1
Name: count, Length: 93, dtype: int64

In [50]:
rest_types = df_zomato['rest_type'].value_counts(ascending= False)
rest_types

rest_type
Quick Bites                   19010
Casual Dining                 10253
Cafe                           3682
Delivery                       2574
Dessert Parlor                 2242
                              ...  
Dessert Parlor, Kiosk             2
Food Court, Beverage Shop         2
Dessert Parlor, Food Court        2
Quick Bites, Kiosk                1
Sweet Shop, Dessert Parlor        1
Name: count, Length: 93, dtype: int64

In [52]:
rest_types_lessthan1000 = rest_types[rest_types<1000]
rest_types_lessthan1000

rest_type
Beverage Shop                 863
Bar                           686
Food Court                    616
Sweet Shop                    468
Bar, Casual Dining            411
                             ... 
Dessert Parlor, Kiosk           2
Food Court, Beverage Shop       2
Dessert Parlor, Food Court      2
Quick Bites, Kiosk              1
Sweet Shop, Dessert Parlor      1
Name: count, Length: 85, dtype: int64

Making Rest Types less than 1000 in frequency as "Others"

In [53]:
def handle_rest_type(value):
    if(value in rest_types_lessthan1000):
        return 'Others'
    else:
        return value
    

In [55]:
df_zomato['rest_type'] = df_zomato['rest_type'].apply(handle_rest_type)
df_zomato['rest_type'].value_counts()

rest_type
Quick Bites           19010
Casual Dining         10253
Others                 9003
Cafe                   3682
Delivery               2574
Dessert Parlor         2242
Takeaway, Delivery     2008
Bakery                 1140
Casual Dining, Bar     1130
Name: count, dtype: int64

Cleaning Location Column

In [56]:
df_zomato['location'].value_counts()

location
BTM                      5056
HSR                      2494
Koramangala 5th Block    2479
JP Nagar                 2218
Whitefield               2105
                         ... 
West Bangalore              6
Yelahanka                   5
Jakkur                      3
Rajarajeshwari Nagar        2
Peenya                      1
Name: count, Length: 93, dtype: int64

There are locations with as high as 5056 restaurants, and others with as low as 1 restaurant. Therefore, I will group locations having less than 300 restaurants in frequency as "others"

In [58]:
location = df_zomato['location'].value_counts(ascending=False)
location_lessthan300 = location[location<300]

In [59]:
def handle_location(value):
    if(value in location_lessthan300):
        return 'others'
    else:
        return value

In [61]:
df_zomato['location'] = df_zomato['location'].apply(handle_location)
df_zomato['location'].value_counts()

location
BTM                      5056
others                   4954
HSR                      2494
Koramangala 5th Block    2479
JP Nagar                 2218
Whitefield               2105
Indiranagar              2026
Jayanagar                1916
Marathahalli             1805
Bannerghatta Road        1609
Bellandur                1268
Electronic City          1246
Koramangala 1st Block    1236
Brigade Road             1210
Koramangala 7th Block    1174
Koramangala 6th Block    1127
Sarjapur Road            1047
Koramangala 4th Block    1017
Ulsoor                   1011
Banashankari              902
MG Road                   893
Kalyan Nagar              841
Richmond Road             803
Malleshwaram              721
Frazer Town               714
Basavanagudi              684
Residency Road            671
Brookefield               656
New BEL Road              644
Banaswadi                 640
Kammanahalli              639
Rajajinagar               591
Church Street             566
L

In [62]:
df_zomato.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

       rest_type                        cuisines  Cost2plates    Type  
0  Casual Dining  North Indian, Mughlai, Chinese          800  Buffet  
1  Casual Dining     Chinese, North Indian, Thai          800  Buffet  
2         Others          Cafe, Mexican, Italian          800  Buffet  
3    Quick Bites      South Indian, North Indian          300  Buffet  
4  Casual Dining        North Indian, Rajasthani          600  Buffet

In [63]:
df_zomato['cuisines'].unique()

array(['North Indian, Mughlai, Chinese', 'Chinese, North Indian, Thai',
       'Cafe, Mexican, Italian', ...,
       'North Indian, Street Food, Biryani', 'Chinese, Mughlai',
       'North Indian, Chinese, Arabian, Momos'], dtype=object)

Cleaning Cuisines Column because the column contains many unique values and few cuisines are having like 1 to 2 restaurants. Therefore, let's group cuisines having less than 100 in frequency as 'others'

In [65]:
cuisines = df_zomato['cuisines'].value_counts(ascending=False)
cuisines_lessthan100 = cuisines[cuisines<100]

In [66]:
def handle_cuisines(value):
    if(value in cuisines_lessthan100):
        return 'others'
    else:
        return value

In [68]:
df_zomato['cuisines'] = df_zomato['cuisines'].apply(handle_cuisines)
df_zomato['cuisines'].value_counts()

cuisines
others                                 26159
North Indian                            2852
North Indian, Chinese                   2351
South Indian                            1820
Biryani                                  903
                                       ...  
South Indian, Chinese, North Indian      105
North Indian, Mughlai, Chinese           104
South Indian, Fast Food                  104
Italian, Pizza                           102
North Indian, Chinese, Seafood           102
Name: count, Length: 70, dtype: int64

In [69]:
df_zomato['Type'].value_counts()

Type
Delivery              25579
Dine-out              17562
Desserts               3559
Cafes                  1703
Drinks & nightlife     1084
Buffet                  869
Pubs and bars           686
Name: count, dtype: int64

The "Type" column has less unique values, therefore, would be left the way it is.

The Data is now Clean. Saving cleaned data to Prepared folder

In [70]:
df_zomato.to_pickle(os.path.join(path, 'Prepared Data', 'zomato_cleaned.pkl'))

In [71]:
df_zomato.to_csv(os.path.join(path, 'Prepared Data', 'zomato_cleaned.csv'))